### ISSUE TITLE
Singularity args not mounted when creating conda environment in singularity+conda usage

### ISSUE THREAD

**Issue by ManavalanG** [snakemake/snakemake#193]
**Snakemake version**
v5.6.0
Based on relevant source code, this code would affect the latest version as well.

**Describe the bug**
When using singularity as well as conda, snakemake doesn't mount `singularity_args`, which results in usage of node's `/tmp` directory instead of mounted `scratch_dir/tmp` directory of choice to Singularity. This fills up node's limited size `/tmp` dir when creating conda env and results in pipeline (with ~10 conda envs) crashing with errors `CreateCondaEnvironmentException` and `NoSpaceLeftError`.

Here I list relevant code that causes this issue.
* Snakemake specifies `$CONDA_PKGS_DIRS` as `/tmp/conda/`:
https://github.com/snakemake/snakemake/blob/2f2556c501987f4e68836c5c82d1d9328ca9207c/snakemake/deployment/conda.py#L324-L325

* Then, when creating conda env using singularity image, it doesn't mount relevant`singularity_args` (unlike in [shell.py](https://github.com/snakemake/snakemake/blob/edda7e0211204d85fa4d02523ddafc2d28919b55/snakemake/shell.py#L117-L125)), where I specify a scratch directory to use as `/tmp/` directory.
https://github.com/snakemake/snakemake/blob/2f2556c501987f4e68836c5c82d1d9328ca9207c/snakemake/deployment/conda.py#L292-L297

**Logs**

```
CreateCondaEnvironmentException:
Could not create conda environment from /scratch/local/tmpypbex5y3.yaml:
Collecting package metadata (repodata.json): ...working... failed

NoSpaceLeftError: No space left on devices.
```

**Example:**

```sh
snakemake --use-conda  --use-singularity  --singularity-args "--bind /data/user/jbond/tmp/:/tmp/"
```

In this example, dir `/tmp` will be used for conda env creation, but dir `/data/user/jbond/tmp` will be used when jobs are run. Ideally, both should use `/data/user/jbond/tmp` in this scenario.

**Additional context**

- This only affects conda env create and not when running jobs as  [shell.py](https://github.com/snakemake/snakemake/blob/edda7e0211204d85fa4d02523ddafc2d28919b55/snakemake/shell.py#L117-L125) does mount `singularity_args`.
- If I manually delete `/tmp/conda`, and then restart snakemake, it can successfully create rest of the conda envs.
- I use SLURM, but this would affect other HPCs as well.

**Comment by ginolhac** [comment]
I have a similar issue with version `5.10.0` that was not present in `5.9.1` when using singularity **without** conda.

```
snakemake --use-singularity --singularity-args ...
```

the additional arguments for `singularity`, _i.e_ the binding of folder, are ignored

**Comment by sschmeier** [comment]
This is a critical. I am having the same problem with 5.10.0. Unable to upgrade.

**Comment by andersgs** [comment]
I am having the same issue, however not related to `--use-conda`. I am still trying to understand completely why it might be happening in my case. Downgrading to 5.5.0 has returned normal functionality for me.

Some debugging effort demonstrates that I am calling `singularity.shell` from within `shell.py`:

```
        if container_img:
            args = context.get("singularity_args", "")
            cmd = singularity.shellcmd(
                container_img,
                cmd,
                args,
                shell_executable=cls._process_args["executable"],
                container_workdir=shadow_dir,
            )
```

But, it seems `singularity_args` is empty. More to the point, it seems the call to context is coming up empty. Suggesting that at some point that is getting overwritten. Not sure yet where.

**Comment by andersgs** [comment]
I should add that it seems to occur if the `Snakefile` is in a separate folder and specified via `-s` flag on the command line.

**Comment by marksantcroos** [comment]
I echo the statement by @ginolhac.

**Comment by andersgs** [comment]
Yes. Me too. I am able to run 5.9.1.

**Comment by marksantcroos** [comment]
Maybe the latter issue deserves it own issue, as the original issue was about something in 5.6.0 and conda.

**Comment by moritzschaefer** [comment]
I've gone a little through it, here is what I found:

For a rule that calls an RScript:

In https://github.com/snakemake/snakemake/blob/53c1589162f296dc509db26235ca6fd232814135/snakemake/script.py#L872
an RScript executor is instantiated. Here the singularity_args argument is still present.

Later, when a shell is spawned to execute singularity, the singularity_args argument is not passed on:
https://github.com/snakemake/snakemake/blob/53c1589162f296dc509db26235ca6fd232814135/snakemake/script.py#L336 (the kwargs argument only contains the script filename)

In shell.py local variables from the script-frame are lifted up, however singularity_args is not defined as local variable there, so it won't get consumed either: https://github.com/snakemake/snakemake/blob/53c1589162f296dc509db26235ca6fd232814135/snakemake/shell.py#L88

My suggested solution would be to either pass singularity_args as part of kwargs in script.py:336 or to just define it locally in script.py such that it gets picked up by the context variable.

**Comment by chrarnold** [comment]
@johanneskoester Please comment upon, it seems to be still there and prevents any updates... Thanks!

**Comment by ginolhac** [comment]
yes, bug still present in version `5.19.3`. Sticking to `5.9.1`

**Comment by johanneskoester** [comment]
Fixed in Snakemake 5.20.1. A PR with some more tests for singularity would be much appreciated, in order to avoid such situations in the future :-).

**Comment by ginolhac** [comment]
bugfix confirmed, all good with `5.20.1` now in bioconda. Thanks again!
